File: runtime-dart/lib/src/draw_batches.dart
part of 'transform.dart';

DrawVertex _vertex(Affine2 world, double lx, double ly, double u, double v) {
  return DrawVertex(
    x: world.a * lx + world.c * ly + world.tx,
    y: world.b * lx + world.d * ly + world.ty,
    u: u,
    v: v,
    r: 1.0,
    g: 1.0,
    b: 1.0,
    a: 1.0,
  );
}

/// Build draw batches for the setup pose, with M7 deformers applied at
/// default parameter values (mirroring the Nim CLI golden-gen pipeline).
///
/// Each slot with a non-empty attachment that resolves to a region becomes one
/// [DrawBatch] with 4 vertices and 6 indices (two triangles).
/// Linear-blend skinning for a mesh attachment's setup vertices, ported fresh to
/// match the Nim `skinMeshVertices` formula and evaluation order
/// (docs/mesh-attachment-contract.md) so both runtimes agree within 1e-4:
///
///   weighted:   worldPos = sum_i weight_i * (boneWorld_i * (bindX_i, bindY_i)),
///               influences accumulated in stored order.
///   unweighted: worldPos = slotBoneWorld * (x, y)  (FK).
///
/// Output x/y/u/v are f32-quantized at the boundary (matching Nim's
/// SkinnedMeshVertex). Meshes carry no per-vertex color in v1, so r=g=b=a=1.
List<DrawVertex> _skinMeshVertices(
  List<Affine2> worlds,
  Map<String, int> boneIndex,
  String slotBone,
  MeshAttachment mesh,
) {
  final out = <DrawVertex>[];
  for (var i = 0; i < mesh.vertices.length; i++) {
    final vertex = mesh.vertices[i];
    final uv = mesh.uvs[i];
    var x = 0.0;
    var y = 0.0;
    if (vertex.weighted) {
      for (final influence in vertex.influences) {
        final w = worlds[boneIndex[influence.bone]!];
        final p = _transformPoint(w, influence.bindX, influence.bindY);
        x += influence.weight * p.x;
        y += influence.weight * p.y;
      }
    } else {
      final p =
          _transformPoint(worlds[boneIndex[slotBone]!], vertex.x, vertex.y);
      x = p.x;
      y = p.y;
    }
    out.add(DrawVertex(
      x: quantizeF32(x),
      y: quantizeF32(y),
      u: quantizeF32(uv.u),
      v: quantizeF32(uv.v),
      r: 1.0,
      g: 1.0,
      b: 1.0,
      a: 1.0,
    ));
  }
  return out;
}

/// Offset skinned mesh [vertices] by a dense per-vertex deform [deltas] list,
/// re-quantizing to f32 at the boundary. Ports `applyDeformDeltas`
/// (runtime-nim/src/bony/mesh/deform.nim:140-153): position is offset, u/v and
/// colour carry through unchanged.
List<DrawVertex> _applyDeformDeltas(
    List<DrawVertex> vertices, List<MeshDelta> deltas) {
  final out = <DrawVertex>[];
  for (var i = 0; i < vertices.length; i++) {
    final v = vertices[i];
    out.add(DrawVertex(
      x: quantizeF32(v.x + deltas[i].x),
      y: quantizeF32(v.y + deltas[i].y),
      u: v.u,
      v: v.v,
      r: v.r,
      g: v.g,
      b: v.b,
      a: v.a,
    ));
  }
  return out;
}

class _DrawBatchBuild {
  const _DrawBatchBuild(
    this.batches,
    this.batchSlotIndex,
    this.batchClipPerTriangle,
  );

  final List<DrawBatch> batches;
  final List<int> batchSlotIndex;
  final List<bool> batchClipPerTriangle;
}

_DrawBatchBuild _buildDrawBatchBuild(
  SkeletonData data,
  List<Affine2> worlds, {
  required String activeSkin,
  required Map<String, SkeletonData> children,
  required bool composeNested,
  required List<String> activeIds,
}) {
  if (worlds.length != data.bones.length) {
    throw FormatException(
        'buildDrawBatches: worlds length ${worlds.length} must match bone count ${data.bones.length}');
  }
  final boneIndex = <String, int>{};
  for (var i = 0; i < data.bones.length; i++) {
    boneIndex[data.bones[i].name] = i;
  }
  final activation = data.activeSkinMembership(activeSkin);
  final attachmentMap = <String, Attachment>{
    for (final attachment in data.allAttachments) attachment.name: attachment,
  };
  // Transient deform-timeline overrides staged on the posed skeleton by
  // applyPose, keyed by slot name + mesh attachment (the mixer produces one
  // entry per slot/attachment).
  final deformMap = <String, List<MeshDelta>>{
    for (final o in data.deformOverrides)
      '${o.slot}\x00${o.attachment}': o.deltas,
  };

  final baseBatches = <DrawBatch>[];
  final batchSlotIndex = <int>[];
  final batchClipPerTriangle = <bool>[];
  final resolvedSlotAttachment = <String, String>{};
  bool meshInfluencesAreActive(MeshAttachment mesh) {
    for (final vertex in mesh.vertices) {
      for (final influence in vertex.influences) {
        final index = boneIndex[influence.bone];
        if (index == null) {
          throw FormatException(
              'mesh influence references unknown bone: ${influence.bone}');
        }
        if (!activation.bones[index]) return false;
      }
    }
    return true;
  }

  for (var slotIdx = 0; slotIdx < data.slots.length; slotIdx++) {
    final slot = data.slots[slotIdx];
    if (slot.attachment.isEmpty) continue;
    final slotBoneIndex = boneIndex[slot.bone]!;
    if (!activation.bones[slotBoneIndex]) continue;
    final attachment = data.resolveSkinAttachmentTarget(
        activeSkin, slot.name, slot.attachment);
    resolvedSlotAttachment[slot.name] = attachment;
    if (attachment.isEmpty) continue;
    switch (attachmentMap[attachment]) {
      case RegionAttachment region:
        final world = worlds[slotBoneIndex];
        final hw = region.width * 0.5;
        final hh = region.height * 0.5;
        baseBatches.add(DrawBatch(
          slot: slot.name,
          bone: slot.bone,
          attachment: attachment,
          blendMode: 'normal',
          texturePage: region.texturePage,
          clipId: '',
          world: world,
          vertices: [
            _vertex(world, -hw, -hh, region.u0, region.v0),
            _vertex(world, hw, -hh, region.u1, region.v0),
            _vertex(world, hw, hh, region.u1, region.v1),
            _vertex(world, -hw, hh, region.u0, region.v1),
          ],
          indices: [0, 1, 2, 2, 3, 0],
        ));
        batchSlotIndex.add(slotIdx);
        batchClipPerTriangle.add(false);
      case MeshAttachment mesh:
        // Skin mesh vertices (FK for unweighted, linear-blend for weighted) and
        // emit one batch in this slot's draw-order position, with metadata
        // mirroring the region path and the Nim reference.
        if (!meshInfluencesAreActive(mesh)) continue;
        final world = worlds[slotBoneIndex];
        var meshVerts = _skinMeshVertices(worlds, boneIndex, slot.bone, mesh);
        // Deform-timeline stage: offset skinned vertices by the posed override
        // for this slot/attachment, immediately after skinning and before the
        // M7 deformer and clipping stages (normative order — see
        // docs/deform-timeline-contract.md).
        final deltas = deformMap['${slot.name}\x00${mesh.name}'];
        if (deltas != null) {
          // Nim's applyDeformDeltas raises schemaViolation on a count mismatch
          // rather than silently rendering the undeformed mesh. Match that: keep
          // the absence guard (no override for this slot/mesh), but a present
          // override whose length disagrees with the skinned vertices is a domain
          // error (defensively unreachable — the loader pins vertexCount ==
          // mesh.vertices.length — but a future invariant break must fail loudly
          // in Dart as it does in Nim, not hide as a static draw).
          if (deltas.length != meshVerts.length) {
            throw FormatException(
                'deform delta count must match skinned vertex count: '
                '${deltas.length} vs ${meshVerts.length}');
          }
          meshVerts = _applyDeformDeltas(meshVerts, deltas);
        }
        baseBatches.add(DrawBatch(
          slot: slot.name,
          bone: slot.bone,
          attachment: attachment,
          blendMode: 'normal',
          texturePage: '',
          clipId: '',
          world: world,
          vertices: meshVerts,
          indices: List<int>.from(mesh.triangles),
        ));
        batchSlotIndex.add(slotIdx);
        batchClipPerTriangle.add(true);
      case NestedRigAttachment nested:
        if (!composeNested) continue;
        if (activeIds.contains(nested.skeleton)) {
          throw FormatException(
              'cycleDetected: nested rig composition cycle detected for skeleton: ${nested.skeleton}');
        }
        final child = children[nested.skeleton];
        if (child == null) {
          throw FormatException(
              'unknownRequiredReference: nested rig child skeleton is not resolved: ${nested.skeleton}');
        }
        final childSkin = nested.skin.isNotEmpty ? nested.skin : 'default';
        if (!child.hasSkin(childSkin)) {
          throw FormatException(
              'unknownRequiredReference: nested rig child skin is not resolved: ${nested.skeleton}/$childSkin');
        }
        final childBuild = _buildDrawBatchBuild(
          child,
          computeWorldTransforms(child, activeSkin: childSkin),
          activeSkin: childSkin,
          children: children,
          composeNested: true,
          activeIds: [...activeIds, nested.skeleton],
        );
        final hostWorld = worlds[slotBoneIndex];
        for (var childIndex = 0;
            childIndex < childBuild.batches.length;
            childIndex++) {
          baseBatches
              .add(_composeBatch(hostWorld, childBuild.batches[childIndex]));
          batchSlotIndex.add(slotIdx);
          batchClipPerTriangle.add(childBuild.batchClipPerTriangle[childIndex]);
        }
      case PathAttachment():
      case PointAttachment():
      case BoundingBoxAttachment():
      case ClippingAttachment():
      case null:
        continue;
    }
  }

  List<DrawBatch> visibleBatches;
  if (data.deformers.isEmpty) {
    visibleBatches = baseBatches;
  } else {
    // Sample each parameter at its default value.
    final samples = data.parameters
        .map((p) => ParameterSample(name: p.name, value: p.defaultValue))
        .toList();
    final efDefs = effectiveDeformers(data.deformers, samples);
    if (efDefs.isEmpty) {
      visibleBatches = baseBatches;
    } else {
      // Apply deformers per batch — each batch uses its own vertices as setup.
      visibleBatches = baseBatches.map((batch) {
        final verts = batch.vertices;
        final positions = verts.map((v) => (x: v.x, y: v.y)).toList();
        final deformed = applyDeformers(positions, efDefs);
        return DrawBatch(
          slot: batch.slot,
          bone: batch.bone,
          attachment: batch.attachment,
          blendMode: batch.blendMode,
          texturePage: batch.texturePage,
          clipId: batch.clipId,
          world: batch.world,
          vertices: [
            for (var i = 0; i < verts.length; i++)
              DrawVertex(
                x: deformed[i].x,
                y: deformed[i].y,
                u: verts[i].u,
                v: verts[i].v,
                r: verts[i].r,
                g: verts[i].g,
                b: verts[i].b,
                a: verts[i].a,
              ),
          ],
          indices: batch.indices,
        );
      }).toList();
    }
  }

  return _DrawBatchBuild(
    _applyClipping(
      data,
      visibleBatches,
      worlds,
      boneIndex,
      resolvedSlotAttachment,
      batchSlotIndex,
      batchClipPerTriangle,
    ),
    batchSlotIndex,
    batchClipPerTriangle,
  );
}

List<DrawBatch> buildDrawBatches(
  SkeletonData data, {
  String activeSkin = 'default',
}) {
  final worlds = computeWorldTransforms(data, activeSkin: activeSkin);
  return _buildDrawBatchBuild(
    data,
    worlds,
    activeSkin: activeSkin,
    children: const {},
    composeNested: false,
    activeIds: const [],
  ).batches;
}

List<DrawBatch> buildNestedDrawBatches(
  SkeletonData data,
  Map<String, SkeletonData> children, {
  String activeSkin = 'default',
  List<Affine2>? worlds,
}) {
  return _buildDrawBatchBuild(
    data,
    worlds ?? computeWorldTransforms(data, activeSkin: activeSkin),
    activeSkin: activeSkin,
    children: children,
    composeNested: true,
    activeIds: const [],
  ).batches;
}

/// Populate `clipId` and geometrically clip covered draw batches, mirroring the
/// Nim reference (`runtime-nim/src/bony/transform.nim`). For each slot whose
/// attachment names a clipping attachment, the covered range is the batch after
/// the clip's own slot through `untilSlot` inclusive (else to the end of draw
/// order). The load-time no-overlap invariant guarantees at most one clip is
/// active per batch. Returns the input unchanged when there are no clips.
List<DrawBatch> _applyClipping(
  SkeletonData data,
  List<DrawBatch> batches,
  List<Affine2> worlds,
  Map<String, int> boneIndex,
  Map<String, String> resolvedSlotAttachment,
  List<int> batchSlotIndex,
  List<bool> batchClipPerTriangle,
) {
  if (data.clippingAttachments.isEmpty) return batches;

  final slotIndexByName = <String, int>{};
  for (var i = 0; i < data.slots.length; i++) {
    slotIndexByName[data.slots[i].name] = i;
  }
  final clipMap = <String, ClippingAttachment>{
    for (final c in data.clippingAttachments) c.name: c,
  };
  final lastSlotIndex = data.slots.length - 1;

  final result = List<DrawBatch>.from(batches);
  for (var slotIdx = 0; slotIdx < data.slots.length; slotIdx++) {
    final slot = data.slots[slotIdx];
    final attachment = resolvedSlotAttachment[slot.name] ?? '';
    if (attachment.isEmpty) continue;
    final clip = clipMap[attachment];
    if (clip == null) continue;
    final ownIndex = slotIdx;
    final endIndex = clip.untilSlot.isNotEmpty
        ? slotIndexByName[clip.untilSlot]!
        : lastSlotIndex;
    // Clip polygon in world space via the clip's own slot's bone world — the
    // same transform the covered region quads are built with.
    final clipWorld = worlds[boneIndex[slot.bone]!];
    final clipPolygon = <ClipPoint>[];
    for (var p = 0; p + 1 < clip.vertices.length; p += 2) {
      final point =
          _transformPoint(clipWorld, clip.vertices[p], clip.vertices[p + 1]);
      clipPolygon.add(ClipPoint(quantizeF32(point.x), quantizeF32(point.y)));
    }
    for (var b = 0; b < result.length; b++) {
      final sourceSlotIndex = batchSlotIndex[b];
      if (sourceSlotIndex <= ownIndex || sourceSlotIndex > endIndex) continue;
      final batch = result[b];
      // Mesh batches clip per-triangle; region batches clip as a convex ring.
      final clipped = batchClipPerTriangle[b]
          ? clipDrawBatchTriangles(batch.vertices, batch.indices, clipPolygon)
          : clipDrawBatchPolygon(batch.vertices, clipPolygon);
      result[b] = DrawBatch(
        slot: batch.slot,
        bone: batch.bone,
        attachment: batch.attachment,
        blendMode: batch.blendMode,
        texturePage: batch.texturePage,
        clipId: clip.name,
        world: batch.world,
        vertices: clipped.changed ? clipped.vertices : batch.vertices,
        indices: clipped.changed ? clipped.indices : batch.indices,
      );
    }
  }
  return result;
}

File: runtime-nim/src/bony/mesh/draw_batches.nim
## Draw-batch assembly and clipping dispatch for transformed skeletons.

import std/tables

import bony/mesh/deform
import bony/mesh/drawbatch_clipping
import bony/mesh/skinning
import bony/model
import bony/transform/affine
import bony/transform/runtime_constraints

type
  NestedSkeletonMap* = Table[string, SkeletonData]

  DrawBatchBuild = object
    batches: seq[DrawBatch]
    batchSlotIndex: seq[int]
    batchClipPerTriangle: seq[bool]

proc vertex(world: Affine2; x, y, u, v: float64): DrawVertex =
  let point = transformPoint(world, x, y)
  DrawVertex(
    x: point.x,
    y: point.y,
    u: u,
    v: v,
    r: 1.0,
    g: 1.0,
    b: 1.0,
    a: 1.0,
  )


proc skinMeshVertices*(
  data: SkeletonData;
  slotBone: string;
  mesh: MeshAttachment;
  skinningMethod = linearBlendSkinning;
): seq[SkinnedMeshVertex] =
  ## Convenience overload relocated from `bony/mesh/skinning` to break the
  ## former transform <-> skinning import cycle: recomputes worlds via the pure
  ## world-transform pass, then defers to the explicit-worlds overload.
  skinMeshVertices(data, computeWorldTransforms(data), slotBone, mesh, skinningMethod)


proc skinMeshVertices*(
  data: SkeletonData;
  slot: SlotData;
  mesh: MeshAttachment;
  skinningMethod = linearBlendSkinning;
): seq[SkinnedMeshVertex] =
  skinMeshVertices(data, slot.bone, mesh, skinningMethod)


proc meshVertex(sv: SkinnedMeshVertex): DrawVertex =
  ## Wrap an already-world-space skinned mesh vertex into a `DrawVertex`. Unlike
  ## `vertex()`, this does NOT re-apply a world transform (skinning already did)
  ## and it carries the uniform region color (r=g=b=a=1): the v1 mesh record has
  ## no per-vertex color, so a mesh and a region on the same slot are
  ## indistinguishable in color.
  DrawVertex(
    x: sv.x,
    y: sv.y,
    u: sv.u,
    v: sv.v,
    r: 1.0,
    g: 1.0,
    b: 1.0,
    a: 1.0,
  )


proc buildDrawBatchBuild(
  data: SkeletonData;
  worlds: seq[Affine2];
  activeSkin: string;
  children: NestedSkeletonMap;
  composeNested: bool;
  activeIds: seq[string];
): DrawBatchBuild =
  ## Internal draw-batch builder. Legacy callers pass `composeNested = false`;
  ## the opt-in nested API passes a host-provided child skeleton map and recurses
  ## through this same path so child clipping happens before parent composition.
  if worlds.len != data.bones.len:
    raise newBonyLoadError(schemaViolation,
      "buildDrawBatches: worlds length " & $worlds.len &
      " must match bone count " & $data.bones.len)
  let activation = activeSkinMembership(data, activeSkin)
  var batches: seq[DrawBatch] = @[]
  var boneIndex = initTable[string, int]()
  var regions = initTable[string, RegionAttachment]()
  var meshes = initTable[string, MeshAttachment]()
  var clips = initTable[string, ClipAttachmentData]()
  var nestedRigs = initTable[string, NestedRigAttachmentData]()
  # Transient deform-timeline override stamped on the posed data by the mixer,
  # keyed by slot name + mesh attachment (the mixer produces one entry per
  # slot/attachment, so keying by slot alone would collapse distinct entries).
  var deformBySlotAttachment = initTable[string, DeformOverride]()

  for index, bone in data.bones:
    boneIndex[bone.name] = index
  for region in data.regions:
    regions[region.name] = region
  for mesh in data.meshAttachments:
    meshes[mesh.name] = mesh
  for clip in data.clippingAttachments:
    clips[clip.name] = clip
  if composeNested:
    for nested in data.nestedRigAttachments:
      nestedRigs[nested.name] = nested
  for override in data.deformOverrides:
    deformBySlotAttachment[override.slot & "\0" & override.attachment] = override

  proc meshInfluencesAreActive(mesh: MeshAttachment): bool =
    for vertex in mesh.vertices:
      for influence in vertex.influences:
        if influence.bone notin boneIndex:
          raise newBonyLoadError(unknownRequiredReference, "mesh influence references unknown bone: " & influence.bone)
        if not activation.bones[boneIndex[influence.bone]]:
          return false
    true

  var slotIndexByName = initTable[string, int]()
  var resolvedSlotAttachment = newSeq[string](data.slots.len)
  for index, slot in data.slots:
    slotIndexByName[slot.name] = index

  # Draw-order slot index of each emitted batch. Batches are a subsequence of the
  # slots (one or more visible batches per drawing slot), so this maps batch ->
  # draw position for computing clip-covered ranges below.
  var batchSlotIndex: seq[int] = @[]
  var batchClipPerTriangle: seq[bool] = @[]
  for slotIdx, slot in data.slots:
    if slot.attachment.len == 0:
      continue
    let slotBoneIndex = boneIndex[slot.bone]
    if not activation.bones[slotBoneIndex]:
      continue
    let attachment = data.resolveSkinAttachmentTarget(activeSkin, slot.name, slot.attachment)
    resolvedSlotAttachment[slotIdx] = attachment
    if attachment.len == 0:
      continue
    if meshes.hasKey(attachment):
      # Keying dispatch on the `meshes` table alone is unambiguous because
      # attachment names are cross-collection unique — validateSkeletonData
      # rejects a mesh name that collides with a region or clip name (model.nim),
      # so a name resolves to at most one of the three tables.
      # Mesh dispatch MUST precede the non-region guard below: a mesh-referencing
      # slot is neither a region nor a clip, so the guard would drop it silently.
      # Skin per-vertex world positions (FK for unweighted, linear-blend for
      # weighted) via the explicit-worlds overload using the worlds we already
      # hold, then emit one batch in this slot's draw-order position. Metadata
      # fields (texturePage/blendMode/clipId/world) mirror the region path so a
      # region and a mesh on the same slot are indistinguishable there.
      let mesh = meshes[attachment]
      if not meshInfluencesAreActive(mesh):
        continue
      # `world` is the slot-bone world used only as batch metadata (mirroring the
      # region path); it does NOT transform the mesh vertices. Skinning consumes
      # the full `worlds` array directly — a weighted vertex blends across its
      # influence bones and ignores slot.bone entirely.
      let world = worlds[boneIndex[slot.bone]]
      var skinned = skinMeshVertices(data, worlds, slot.bone, mesh)
      # Deform-timeline stage: offset skinned vertices by the posed override for
      # this slot/attachment, immediately after skinning and before the M7
      # deformer and clipping stages (normative order — see
      # docs/deform-timeline-contract.md). Region batches are never offset.
      let deformKey = slot.name & "\0" & mesh.name
      if deformKey in deformBySlotAttachment:
        let override = deformBySlotAttachment[deformKey]
        if override.deltas.len == skinned.len:
          skinned = applyDeformDeltas(skinned, override.deltas)
      var meshVerts = newSeq[DrawVertex](skinned.len)
      for i, sv in skinned:
        meshVerts[i] = meshVertex(sv)
      batches.add DrawBatch(
        slot: slot.name,
        bone: slot.bone,
        attachment: attachment,
        texturePage: "",
        blendMode: "normal",
        clipId: "",
        world: world,
        vertices: meshVerts,
        indices: mesh.triangles,
      )
      batchSlotIndex.add slotIdx
      batchClipPerTriangle.add true
      continue
    if composeNested and nestedRigs.hasKey(attachment):
      let nested = nestedRigs[attachment]
      if nested.skeleton in activeIds:
        raise newBonyLoadError(cycleDetected,
          "nested rig composition cycle detected for skeleton: " & nested.skeleton)
      if not children.hasKey(nested.skeleton):
        raise newBonyLoadError(unknownRequiredReference,
          "nested rig child skeleton is not resolved: " & nested.skeleton)
      let childSkin = if nested.skin.len > 0: nested.skin else: "default"
      let child = children[nested.skeleton]
      if not child.hasSkin(childSkin):
        raise newBonyLoadError(unknownRequiredReference,
          "nested rig child skin is not resolved: " & nested.skeleton & "/" & childSkin)
      var nextIds = activeIds
      nextIds.add nested.skeleton
      let hostWorld = worlds[boneIndex[slot.bone]]
      let childBuild = buildDrawBatchBuild(
        child,
        computeWorldTransforms(child, childSkin),
        childSkin,
        children,
        true,
        nextIds,
      )
      for childIndex, childBatch in childBuild.batches:
        batches.add composeBatch(hostWorld, childBatch)
        batchSlotIndex.add slotIdx
        batchClipPerTriangle.add childBuild.batchClipPerTriangle[childIndex]
      continue
    if not regions.hasKey(attachment):
      # A slot whose attachment names a clipping attachment (or any non-region)
      # produces no draw batch. This guard also prevents the `regions[...]`
      # Table lookup below from raising KeyError on a clip-named slot.
      continue
    let region = regions[attachment]
    let index = boneIndex[slot.bone]
    let world = worlds[index]
    let halfWidth = region.width * 0.5
    let halfHeight = region.height * 0.5
    batches.add DrawBatch(
      slot: slot.name,
      bone: slot.bone,
      attachment: attachment,
      texturePage: region.texturePage,
      blendMode: "normal",
      clipId: "",
      world: world,
      vertices: @[
        vertex(world, -halfWidth, -halfHeight, region.u0, region.v0),
        vertex(world, halfWidth, -halfHeight, region.u1, region.v0),
        vertex(world, halfWidth, halfHeight, region.u1, region.v1),
        vertex(world, -halfWidth, halfHeight, region.u0, region.v1),
      ],
      indices: @[0'u16, 1'u16, 2'u16, 2'u16, 3'u16, 0'u16],
    )
    batchSlotIndex.add slotIdx
    batchClipPerTriangle.add false

  # Clip pass: each slot whose attachment names a clipping attachment sets
  # `clipId` on, and geometrically clips, the draw batches in its covered range
  # (the batch after the clip's own slot through `untilSlot` inclusive, else to
  # the end of draw order). The load-time no-overlap invariant guarantees at most
  # one clip is active over any batch, so clips are applied independently.
  if data.clippingAttachments.len > 0:
    let lastSlotIndex = data.slots.len - 1
    for slotIdx, slot in data.slots:
      let slotBoneIndex = boneIndex[slot.bone]
      if not activation.bones[slotBoneIndex]:
        continue
      let attachment = resolvedSlotAttachment[slotIdx]
      if attachment.len == 0 or not clips.hasKey(attachment):
        continue
      let clip = clips[attachment]
      let ownIndex = slotIdx
      let endIndex =
        if clip.untilSlot.len > 0: slotIndexByName[clip.untilSlot]
        else: lastSlotIndex
      # Clip polygon in world space via the clip's own slot's bone world — the
      # same transform the covered region quads are built with.
      let clipWorld = worlds[slotBoneIndex]
      var clipPolygon: seq[ClipPoint] = @[]
      var p = 0
      while p + 1 < clip.vertices.len:
        let point = transformPoint(clipWorld, clip.vertices[p], clip.vertices[p + 1])
        clipPolygon.add clipPoint(
          quantizeF32(point.x, "clip.poly.x"),
          quantizeF32(point.y, "clip.poly.y"),
        )
        p += 2
      for batchIdx in 0 ..< batches.len:
        let sourceSlotIndex = batchSlotIndex[batchIdx]
        if sourceSlotIndex <= ownIndex or sourceSlotIndex > endIndex:
          continue
        if batchClipPerTriangle[batchIdx]:
          # Meshes are a triangle *soup* (explicit index list, shared/interior
          # vertices), so they clip per-triangle via clipDrawBatchTriangles —
          # NOT through clipDrawBatchPolygon, which reinterprets a batch's
          # vertices as one convex boundary ring and would destroy the topology.
          # See docs/mesh-attachment-contract.md ("per-triangle mesh clipping").
          batches[batchIdx].clipId = clip.name
          let clipped = clipDrawBatchTriangles(
            batches[batchIdx].vertices, batches[batchIdx].indices, clipPolygon)
          if clipped.changed:
            batches[batchIdx].vertices = clipped.vertices
            batches[batchIdx].indices = clipped.indices
          continue
        batches[batchIdx].clipId = clip.name
        let clipped = clipDrawBatchPolygon(batches[batchIdx].vertices, clipPolygon)
        if clipped.changed:
          batches[batchIdx].vertices = clipped.vertices
          batches[batchIdx].indices = clipped.indices

  DrawBatchBuild(
    batches: batches,
    batchSlotIndex: batchSlotIndex,
    batchClipPerTriangle: batchClipPerTriangle,
  )


proc buildDrawBatches*(data: SkeletonData; worlds: seq[Affine2]; activeSkin = "default"): seq[DrawBatch] =
  ## Build draw batches using caller-supplied world transforms. Callers that have
  ## advanced the stateful physics stage pass the physics-adjusted worlds here so
  ## draw-batch vertices reflect physics. `worlds[i]` must be the world transform
  ## of `data.bones[i]` — same length and index ordering as `computeWorldTransforms`.
  ## Raised (not `doAssert`ed) so the guard survives `-d:danger`: a stripped check
  ## here would reintroduce the silent out-of-bounds read this seam exists to prevent.
  var noChildren = initTable[string, SkeletonData]()
  buildDrawBatchBuild(data, worlds, activeSkin, noChildren, false, @[]).batches


proc buildNestedDrawBatches*(
  data: SkeletonData;
  worlds: seq[Affine2];
  children: NestedSkeletonMap;
  activeSkin = "default";
): seq[DrawBatch] =
  ## Opt-in host-resolved nested rig setup-pose composition. `children` maps each
  ## `NestedRigAttachmentData.skeleton` id to an already-loaded child
  ## `SkeletonData`. Legacy `buildDrawBatches` remains non-composing.
  buildDrawBatchBuild(data, worlds, activeSkin, children, true, @[]).batches


proc buildDrawBatches*(data: SkeletonData): seq[DrawBatch] =
  ## Convenience overload: recompute worlds from the pure world-transform pass.
  ## Use the `worlds` overload when a physics stage has adjusted bone worlds.
  buildDrawBatches(data, computeWorldTransforms(data))


proc buildDrawBatches*(data: SkeletonData; activeSkin: string): seq[DrawBatch] =
  ## Convenience overload for runtime skin selection with pure world transforms.
  buildDrawBatches(data, computeWorldTransforms(data, activeSkin), activeSkin)


proc buildNestedDrawBatches*(
  data: SkeletonData;
  children: NestedSkeletonMap;
  activeSkin = "default";
): seq[DrawBatch] =
  ## Convenience overload for setup-pose nested composition with pure world
  ## transforms.
  buildNestedDrawBatches(data, computeWorldTransforms(data, activeSkin), children, activeSkin)
